File: services/analyzer/analyzers/balance.py
"""Balance analyzer: visual weight centroid versus the geometric canvas
center. Symmetric/centered designs can be intentional -- we only flag a
large, likely-unintentional offset.

Weight is |luminance - background luminance|, not gradient magnitude. Pure
edge/gradient weighting only registers the *boundary* of a filled shape and
misses a large solid block's interior almost entirely -- exactly the kind of
element that visually dominates a layout. Deviation-from-background is a
much closer proxy for "how much this pixel pulls the eye"."""

from __future__ import annotations

import cv2
import numpy as np

from models import Finding, Measured
from shared import TextRegion, new_id

EXPECTED_OFFSET_FRAC = (0.0, 0.2)  # offset / half-diagonal


def analyze(bgr: np.ndarray, _text_regions: list[TextRegion]) -> list[Finding]:
    h, w = bgr.shape[:2]
    gray = cv2.cvtColor(bgr, cv2.COLOR_BGR2GRAY).astype(np.float64)

    # Median is a robust estimate of the dominant background tone (most of a
    # poster/design's area is background, even a busy one).
    background = float(np.median(gray))
    weight = np.abs(gray - background)

    total = weight.sum()
    if total <= 0:
        return []

    ys, xs = np.indices(weight.shape)
    cx = float((xs * weight).sum() / total)
    cy = float((ys * weight).sum() / total)

    center_x, center_y = w / 2, h / 2
    offset = float(np.hypot(cx - center_x, cy - center_y))
    half_diag = float(np.hypot(w, h) / 2)
    offset_frac = offset / half_diag if half_diag > 0 else 0.0

    lo, hi = EXPECTED_OFFSET_FRAC
    if offset_frac <= hi:
        return []

    box_size = min(w, h) * 0.15
    bbox = (
        max(0.0, cx - box_size / 2),
        max(0.0, cy - box_size / 2),
        box_size,
        box_size,
    )

    return [
        Finding(
            id=new_id(),
            dimension="balance",
            severity="major" if offset_frac > hi * 1.5 else "minor",
            bbox=bbox,
            measured=Measured(value=round(offset_frac, 2), expected=(lo, hi), unit=" centroid offset"),
        )
    ]

File: src/lib/measure/analyzers/balance.ts
import { newFindingId } from "../id";
import { SKIPPED, measured, type AnalyzerResult } from "./_result";

/**
 * Balance analyzer: visual weight centroid versus the geometric canvas
 * center. Symmetric/centered designs can be intentional — only a large,
 * likely-unintentional offset is flagged. Weight is
 * |luminance - background luminance|, not gradient magnitude, since a
 * large solid block's interior (exactly the kind of element that visually
 * dominates a layout) has almost no edge signal but a strong deviation from
 * the background. Port of `services/analyzer/analyzers/balance.py`.
 */
const EXPECTED_OFFSET_FRAC: [number, number] = [0.0, 0.2];

export function analyzeBalance(
  gray: ArrayLike<number>,
  width: number,
  height: number,
): AnalyzerResult {
  const sorted = Float32Array.from(gray).sort();
  const background = sorted[Math.floor(sorted.length / 2)];

  let total = 0;
  let sumX = 0;
  let sumY = 0;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const w = Math.abs(gray[y * width + x] - background);
      total += w;
      sumX += x * w;
      sumY += y * w;
    }
  }
  if (total <= 0) return SKIPPED;

  const cx = sumX / total;
  const cy = sumY / total;
  const centerX = width / 2;
  const centerY = height / 2;
  const offset = Math.hypot(cx - centerX, cy - centerY);
  const halfDiag = Math.hypot(width, height) / 2;
  const offsetFrac = halfDiag > 0 ? offset / halfDiag : 0;

  const [lo, hi] = EXPECTED_OFFSET_FRAC;
  if (offsetFrac <= hi) return measured();

  const boxSize = Math.min(width, height) * 0.15;
  const bbox: [number, number, number, number] = [
    Math.max(0, cx - boxSize / 2),
    Math.max(0, cy - boxSize / 2),
    boxSize,
    boxSize,
  ];

  return measured([
    {
      id: newFindingId(),
      dimension: "balance",
      severity: offsetFrac > hi * 1.5 ? "major" : "minor",
      bbox,
      measured: { value: Math.round(offsetFrac * 100) / 100, expected: [lo, hi], unit: " centroid offset" },
    },
  ]);
}
